### DISCUSSION TITLE
How to identify if a page has an image or not

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#3848] [Digital Image Processing]
I have pages that have

1. Plain white background with text
2. Color background with text
3. plain white background with image
4. Color background with image
5. color background with image and text

How do I identify if the page has an image or not?

Thanks,
![Uploading abc.jpg…]()
![Uploading xyz.jpg…]()

**Comment by snibgo** [comment]
You don't show any sample inputs. These would help.

What is a "page"? Do you mean "page of a PDF file"?

What version of IM? On what platform?

**Reply by srini8080** [reply]
![abc](https://user-images.githubusercontent.com/12046302/122692278-b1141780-d251-11eb-94bc-048fd0d551f0.jpg)
![xyz](https://user-images.githubusercontent.com/12046302/122692339-fafcfd80-d251-11eb-82fc-cafa582461e2.jpg)

Yes, page of a pdf file but can be any format. JPG/PNG/GIF/PDF/PPM

IM 6.9.10-71 Q16 

Linux.

Thanks

**Comment by snibgo** [comment]
PDF files can contain vectors as well as raster images. Specialist tools can be used for PDF. For example, the pdfimages program can tell you whether the PDF contains any raster images, and their sizes, and so on.

For raster images:

1. If the overall standard deviation is zero, the image is one solid colour.

2. If the image contains a photograph, then the photo might contain text, such as a street sign. Identifying these with high accuracy is a very complex process.

3. If text is written on a plain background, it usually has a high local standard deviation, where "local" means an area roughly twice the width of the character strokes. For example, use "-statistic StandardDeviation 3x3". However, line drawings also have this characteristic. So this can be used as a negative test ("the image has neither text nor line drawings"), but not a positive test for text or line drawings in isolation.

4. Morphology can be used to smear the image horizontally, by blur or open or close. Then connected components can be used. Any components that are much wider than they are tall may be text. But components that are both wide and tall are probably illustrations.

Using these methods, it is possible to partition a scanned image into areas that are (a) blank paper or (b) text written in horizontal rows or (c) illustrations. But the process is complex, and I don't have an easy solution.

**Comment by srini8080** [comment]
The problem is I have scanned images of a book that show the entire page as an image in the pdf.

I tried connected components method but the error rate of identification is huge.

Is there a way we can just say that the image (page) has something other than text (ofcourse other than color background)

**Reply by fmw42** [reply]
Use tesseract to read text. What is left is not text.

**Reply by srini8080** [reply]
I tried that. A page has text and an image. Tesseract is showing that there is text. How do I know that there is an image?

**Comment by srini8080** [comment]
I saw this post from a decade ago. 

The multi_crop does not work as the entire page is showing up as an image.

The python code is giving out good indication of if there is a image. How do I make use of it?

from PIL import Image
 
im = Image.open('D:/pavian.png', 'r')
w, h = im.size
a = [[0]*w for i in range(h)]
b = [[0]*w for i in range(h)]
for i in range(h):
    for j in range(w):
        a[i][j] = im.getpixel((j, i))

d = [[-1,-1], [-1,0], [-1,1], [0,1], [1,1], [1,0], [1,-1], [0,-1]]
 
c = 40    # // порог разницы в интенсивностях двух соседей
s = 20    # // сторона квадратиков
 
def foo(p, q):
    cnt = 0
    for i in range(p + 1, p + s - 1):
        for j in range(q + 1, q + s - 1):
            for k in range(8):
                if abs(a[i][j] - a[i + d[k][0]][j + d[k][1]] > c):
                    cnt += 1
    return cnt

z = 0
for i in range(0, h - s, s):
    for j in range(0, w - s, s):
        p = foo(i, j)
        for k in range(s):
            for l in range(s):
                b[i + k][j + l] = p
        z = max(z, p)
 
 
for i in range(h):
    for j in range(w):
        if b[i][j] > z / 2:
            v = 255
        else:
            v = 0
        im.putpixel((j, i), v)
 
f_out = open('D:/pavian_out.png', 'wb')
im.save(f_out)
f_out.close()

Thanks

**Comment by snibgo** [comment]
In python, spaces at the start of lines are significant. But the forum software has removed them.

When posting code, please put a line with three back-ticks before the code, and the same after the code.

**Reply by srini8080** [reply]
Sorry. A little new to this.

Here is the code that seems to help. Wanted to know how I can use this to identify if an image has a picture inside?

```

from PIL import Image
import os, sys


try:
    filename = sys.argv[1]
except IndexError:
    print ("You did not specify a image file")
    sys.exit(1)

#filename.lower().endswith(('.png', '.jpg', '.jpeg', '.tiff', '.bmp', '.gif'))
if not filename.lower().endswith('.png'):
    print("Image file has to be a png")
    sys.exit(1)

base=os.path.basename(filename)
filebase=os.path.splitext(base)[0]
fileext=os.path.splitext(base)[0]

im = Image.open(filename, 'r')
w, h = im.size
a = [[0]*w for i in range(h)]
b = [[0]*w for i in range(h)]
for i in range(h):
    for j in range(w):
        a[i][j] = im.getpixel((j, i))
 
 
d = [[-1,-1], [-1,0], [-1,1], [0,1], [1,1], [1,0], [1,-1], [0,-1]]
 
 
c = 40    # // порог разницы в интенсивностях двух соседей
s = 20    # // сторона квадратиков
 
 
def foo(p, q):
    cnt = 0
    for i in range(p + 1, p + s - 1):
        for j in range(q + 1, q + s - 1):
            for k in range(8):
                if abs(a[i][j] - a[i + d[k][0]][j + d[k][1]] > c):
                    cnt += 1
    return cnt
 
 
z = 0
for i in range(0, h - s, s):
    for j in range(0, w - s, s):
        p = foo(i, j)
        for k in range(s):
            for l in range(s):
                b[i + k][j + l] = p
        z = max(z, p)
 
 
for i in range(h):
    for j in range(w):
        if b[i][j] > z / 2:
            v = 255
        else:
            v = 0
        im.putpixel((j, i), v)
 
f_out = open(filebase+"_out"+".png", 'wb')
im.save(f_out)
f_out.close()

```

**Reply by fmw42** [reply]
Please note that this is an ImageMagick forum, not a Python forum.

**Reply by srini8080** [reply]
Sorry, did not realize that my post did not have the link earlier

That code was picked up here...

https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=18686

**Reply by fmw42** [reply]
The code on that page is about finding groups of 1s and 0s and does not identify whether the image has text or images or its content. You have also added quite a bit of code to what end?  Are you asking to have this converted to ImageMagick? If so, then it needs coding and is not something that exists and not likely to be implemented in the near term as the developers are very busy.  

Personally, I do not see how you can reliably identify what the content of the image might contain in the way of images vs graphics vs text.  You can use tesseract to read text and perhaps find where it exists. The rest would be likely images or graphics.  Text comes in regular rows. Images and graphics are generally larger. Graphics usually has lots of constant regions of colors, but may be more complex as well. Images, typically have lots of texture.

Perhaps you could train artificial intelligence to recognize the presence of imagery. But I have not seen any papers on that topic, though I have not researched the topic.
